lib/DBIx/Class/ResultSource.pm

```diff
@@ -874,7 +889,41 @@ column names as values.
 =cut
 
 sub unique_constraints {
-  return %{shift->_unique_constraints||{}};
+  my $uniques = shift->_unique_constraints || {};
+
+  return map { $_ => $uniques->{$_}->{columns} } keys %$uniques;
+}
+
+=head2 unique_constraints_info
+
+=over 4
+
+=item Arguments: none
+
+=item Return Value: Hashref of unique constraint data
+
+=back
+
+  my $unique_info = $source->unique_constraints_info();
+
+Read-only accessor returning all information about unique constraints.
+
+The hashref is keyed by the constraint name, and the values are
+hashrefs contaning two keys:
+
+C<columns> is an arrayref of the columns originally provided to
+L</add_unique_constraint> or L</add_unique_constraints>;
+
+C<extra> is the hashref originally passed to
+L</add_unique_constraint>, or an empty hashref if none was provided.
```

Comment:
Do not do `an empty hashref if none was provided`. Either the user supplied one or they did not, do not add anything on your own.